subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildSessionScopeServices.java

```diff
@@ -165,7 +165,11 @@ ScriptSourceHasher createScriptSourceHasher(FileHasher fileHasher) {
     }
 
     FileSystemSnapshotter createFileSystemSnapshotter(FileHasher hasher, StringInterner stringInterner, FileSystem fileSystem, FileSystemMirror fileSystemMirror) {
-        return new DefaultFileSystemSnapshotter(hasher, stringInterner, fileSystem, fileSystemMirror, DirectoryScanner.getDefaultExcludes());
+        // added to fix  MD5 error when .gradle project cache is part of the root source code dirs (typically in native builds)
+        // issue: https://github.com/gradle/gradle/issues/5941
+        DirectoryScanner.addDefaultExclude("**/.gradle/**"); // should never include the .gradle directory in a snapshot
```

Comment:
According to the issue, we should use `ProjectCacheDir#getDir()` instead of hardcoding `.gradle`.

Reply:
Like it is done here, `.gradle` directories will not be included in any snapshot, nor will they be visited when querying any file collection, given that the `DirectoryWalker` also respects default excludes (see [here](https://github.com/gradle/gradle/blob/f60286b0fd331972b26a392628b79615ae2a1138/subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java#L81-L92)).

While I am not against adding `.gradle` as a default exclude, I don't think this is the right place here.
I think the change to the default excludes should be done once for each build. Maybe just before we create the `DefaultGradle` instance: https://github.com/gradle/gradle/blob/946a56b100aaf4901e9926c9b631dcb1bef7a3ec/subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java#L176